### DISCUSSION TITLE
PM2-runtime with max instance.

### DISCUSSION THREAD

**Discussion by iamrehank** [vercel/next.js#64702] [Help]
### Summary

Can anyone guide me to configure PM2 with docker. 
I have tried it but it not working for me. below are file configuration. 

DockerFile :
FROM node:18-alpine

WORKDIR /app

RUN npm install --global pm2

COPY . /app
RUN npm install --f

RUN npm run build

EXPOSE 3000

CMD ["pm2-runtime", "ecosystem.config.js"]

ecosystem.config.js

module.exports = {
  apps: [
    {
      name: "nextjs-app",
      script: "npm",
      args: "run start",
      instances: "max",
      autorestart: false,
      watch: false
    },
  ],
};

Command : pm2-runtime start ecosystem.config.js

**Comment by SouthLink** [comment]
I got my pm2 next.js program to run in docker with the following configuration, which works for me.

my dockerfile
```
FROM node:16-slim as next-base

########               install + builder              ########
FROM next-base AS installBuilder

# Set work path
WORKDIR /www

# Copy package.json
COPY ./package.json /www/package.json
COPY ./.npmrc /www/.npmrc
COPY ./pnpm-lock.yaml /www/pnpm-lock.yaml 

RUN  npm install webpack -g && npm install -g pnpm@8.7.5 && pnpm install

# Copy package.json
COPY . /www

RUN pnpm run build:test  && rm -rf /www/.next/cache

########              runner                ########
FROM next-base AS runner

RUN groupadd www -g8080
RUN useradd www -u8080 -g8080

WORKDIR /www

RUN apt-get update && apt-get install nginx git -y

# pm2 
RUN npm install pm2@latest -g
RUN pm2 update
RUN pm2 install pm2-intercom

COPY --from=installBuilder /www ./

COPY ./_build/nginx_develop.conf /etc/nginx/nginx.conf

CMD pm2 start ecosystem.config.js --env production && nginx -g 'daemon off;'

```

my ecosystem.config.js

```
const path = require('path');
const fs = require('fs');

const BUILD_ID = fs.readFileSync(path.join(__dirname, '.next', 'BUILD_ID'), 'utf-8');

module.exports = {
  apps: [{
    name: "my-app",
    script: "server.js",
    instances: 5,
    instance_var: "INSTANCE_ID",
    exec_mode: "cluster",
    env: {
      NODE_ENV: "development",
      SERVER_ENV: "development",
      BUILD_ID: BUILD_ID
    },
    env_test: {
      NODE_ENV: "test",
      SERVER_ENV: "test",
      BUILD_ID: BUILD_ID
    },
    env_production: {
      NODE_ENV: "production",
      SERVER_ENV: "production",
      BUILD_ID: BUILD_ID
    },
    max_memory_restart: "700M"
  }]
}

```

I use a custom server. If you want to use standalone mode, you can use this example [with-docker](https://github.com/vercel/next.js/tree/canary/examples/with-docker) and my dockerfile to modify it

**Reply by kachmul2004** [reply]
could you please share your nginx.conf settings? I have issues getting node to run because it tries to direct traffic to an ipv6 address

**Reply by SouthLink** [reply]
If you don't explicitly specify an IPv4 address in the listen directive, Nginx may try to use an IPv6 address by default. Make sure you specify an IPv4 address explicitly in the listen directive, such as listen 80; Port 80 of the IPv4 address is monitored.
You need to make sure that the proxy configuration uses IPv4 addresses rather than IPv6 addresses
```
user www www;
worker_processes  10;
error_log  /dev/stderr;

events {
    use epoll;
    worker_connections  10240;
}

http {
    include       mime.types;
    default_type  application/octet-stream;

    access_log  /dev/stdout;

    server_names_hash_bucket_size    128;
    client_header_buffer_size        32k;
    large_client_header_buffers    4 32k;
    client_max_body_size              8m;

    sendfile       on;
    tcp_nopush     on;
    tcp_nodelay    on;

    keepalive_timeout  30;

    gzip                   on;
    gzip_min_length        1k;
    gzip_buffers        4 16k;
    gzip_http_version     1.0;
    gzip_comp_level         2;
    gzip_vary              on;
    gzip_types          text/plain application/x-javascript text/css application/xml;
    

    #error_page 403 404 500 502 503 504  @failback;

    upstream my-next {
        server 127.0.0.1:3000;
        keepalive 256;
    }

    server {
        listen 80;

        location / {
            proxy_pass http://my-next;
        }
        
        location /_next/ {
            alias /www/.next/;
        }
    }
}
```

**Reply by kachmul2004** [reply]
> If you don't explicitly specify an IPv4 address in the listen directive, Nginx may try to use an IPv6 address by default. Make sure you specify an IPv4 address explicitly in the listen directive, such as listen 80; Port 80 of the IPv4 address is monitored.
> You need to make sure that the proxy configuration uses IPv4 addresses rather than IPv6 addresses
> ```
> user www www;
> worker_processes  10;
> error_log  /dev/stderr;
> 
> events {
>     use epoll;
>     worker_connections  10240;
> }
> 
> http {
>     include       mime.types;
>     default_type  application/octet-stream;
> 
>     access_log  /dev/stdout;
> 
>     server_names_hash_bucket_size    128;
>     client_header_buffer_size        32k;
>     large_client_header_buffers    4 32k;
>     client_max_body_size              8m;
> 
>     sendfile       on;
>     tcp_nopush     on;
>     tcp_nodelay    on;
> 
>     keepalive_timeout  30;
> 
>     gzip                   on;
>     gzip_min_length        1k;
>     gzip_buffers        4 16k;
>     gzip_http_version     1.0;
>     gzip_comp_level         2;
>     gzip_vary              on;
>     gzip_types          text/plain application/x-javascript text/css application/xml;
>     
> 
>     #error_page 403 404 500 502 503 504  @failback;
> 
>     upstream my-next {
>         server 127.0.0.1:3000;
>         keepalive 256;
>     }
> 
>     server {
>         listen 80;
> 
>         location / {
>             proxy_pass http://my-next;
>         }
>         
>         location /_next/ {
>             alias /www/.next/;
>         }
>     }
> }
> ```

Thanks. Problem solved!

**Comment by iamrehank** [comment]
I'm trying to use pm2-runtime as described in the PM2 documentation.
